src/transaction/transaction.rs

```diff
@@ -305,15 +322,42 @@ impl Transaction {
     /// # });
     /// ```
     pub async fn commit(&mut self) -> Result<()> {
-        TwoPhaseCommitter::new(
+        self.check_status()?;
+        let res = TwoPhaseCommitter::new(
             self.buffer.to_proto_mutations().await,
             self.timestamp.version(),
             self.bg_worker.clone(),
             self.rpc.clone(),
             self.for_update_ts,
         )
         .commit()
-        .await
+        .await;
+
+        if res.is_ok() {
+            self.status = TransactionStatus::Committed;
+        }
+        res
+    }
+
+    /// Rollback the transaction.
+    ///
+    /// If it succeeds, all mutations made by this transaciton will not take effect.
+    pub async fn rollback(&mut self) -> Result<()> {
+        self.check_status()?;
```

Comment:
Can we check to see if we need to rollback? If we haven't taken any locks or done select for update, then I think we can just drop the transaction

Reply:
The check is in performed in `TwoPhaseCommitter`.